### PR TITLE
Support two word AddCommands

### LINKED ISSUE #78: Support 2 words Add Command

### PR DESCRIPTION
**Pull request by KengHian**
- close #78 
- Implemented template`<T>` in `AddCommand` as discussed this afternoon

### CODE REVIEW

File: src/main/java/seedu/address/logic/commands/AddContactCommand.java [hunk 1]
```diff
@@ -7,16 +7,15 @@
...
 /**
  * Adds a person to the address book.
  */
-public class AddContactCommand extends Command {
-    public static final String COMMAND_WORD = "add_contact";
+public class AddContactCommand extends AddCommand<Person> {
+    public static final String COMMAND_ENTITY = "contact";
 
     public static final String MESSAGE_USAGE = COMMAND_WORD + ": Adds a contact to the address book. "
             + "Parameters: "
```
Comment (Green-Tea-123) [inline src/main/java/seedu/address/logic/commands/AddContactCommand.java:21]: I think you might have missed updating the MESSAGE_USAGE to COMMAND_WORD + " " + COMMAND_ENTITY

File: src/main/java/seedu/address/logic/commands/AddCompanyCommand.java [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+/**
+ * Adds a company to the address book.
+ */
+public class AddCompanyCommand extends AddCommand<Company> {
+
+    public static final String COMMAND_ENTITY = "company";
+
+    public static final String MESSAGE_USAGE =
+            COMMAND_WORD + ": Adds a company to the address book. ";
```
Comment (Green-Tea-123) [inline src/main/java/seedu/address/logic/commands/AddCompanyCommand.java:16]: Same issue as AddContactCommand

File: src/main/java/seedu/address/logic/commands/AddJobCommand.java [hunk 3]
```diff
@@ -7,45 +7,33 @@
...
-            + PREFIX_SALARY + "SALARY "
-            + "[" + PREFIX_REQUIREMENTS + "REQUIREMENTS] "
-            + "[" + PREFIX_DESCRIPTION + "DESCRIPTION]\n"
-            + "Example: " + COMMAND_WORD + " "
-            + PREFIX_NAME + "Waiter "
-            + PREFIX_COMPANY + "Starbucks "
-            + PREFIX_SALARY + "2500 "
-            + PREFIX_REQUIREMENTS + "Hands and feet "
+    public static final String MESSAGE_USAGE =
+            COMMAND_WORD + ": Adds a job listing to the address book. " + "Parameters: " + PREFIX_NAME + "JOBNAME "
```
Comment (Green-Tea-123) [inline src/main/java/seedu/address/logic/commands/AddJobCommand.java:23]: Same issue as AddContactCommand

File: src/main/java/seedu/address/logic/parser/AddCommandParser.java [hunk 4]
```diff
@@ -1,61 +1,47 @@
...
  */
-public class AddCommandParser implements Parser<AddContactCommand> {
+public class AddCommandParser implements Parser<AddCommand<?>> {
+
+    /**
+     * Returns true if none of the prefixes contains empty {@code Optional} values in the given
+     * {@code ArgumentMultimap}.
+     */
+    private static boolean arePrefixesPresent(ArgumentMultimap argumentMultimap, Prefix... prefixes) {
+        return Stream.of(prefixes).allMatch(prefix -> argumentMultimap.getValue(prefix).isPresent());
```
Comment (Green-Tea-123) [inline src/main/java/seedu/address/logic/parser/AddCommandParser.java:21]: I don't think this function is being used at all, it can probably be removed.

**Review by Green-Tea-123 — APPROVED** [review]
Some small issues, otherwise LGTM.

File: src/main/java/seedu/address/commons/exceptions/NotImplementedException.java [hunk 5]
```diff
@@ -0,0 +1,7 @@
+package seedu.address.commons.exceptions;
+
+/**
+ * Placeholder when a feature is not implemented yet.
+ */
+public class NotImplementedException extends RuntimeException {
+}
```
[the remark this answers is not in the document]
Comment (KengHian) [reply src/main/java/seedu/address/commons/exceptions/NotImplementedException.java:7]: Good suggestion, will add a comment here!

File: src/main/java/seedu/address/logic/commands/AddContactCommand.java (hunk 1 above)
Comment (KengHian) [reply src/main/java/seedu/address/logic/commands/AddContactCommand.java:21]: Yup good catch

File: src/main/java/seedu/address/logic/parser/AddCommandParser.java (hunk 4 above)
Comment (KengHian) [reply src/main/java/seedu/address/logic/parser/AddCommandParser.java:21]: Yeah true, will do!

File: src/main/java/seedu/address/logic/commands/AddJobCommand.java [hunk 6]
```diff
@@ -16,28 +15,32 @@
...
 
-    public static final String COMMAND_WORD = "add_job";
-
-    public static final String MESSAGE_USAGE =
-            COMMAND_WORD + ": Adds a job listing to the address book. " + "Parameters: " + PREFIX_NAME + "JOBNAME "
-            + PREFIX_COMPANY + "COMPANY " + PREFIX_SALARY + "SALARY " + "[" + PREFIX_REQUIREMENTS + "REQUIREMENTS] "
-            + "[" + PREFIX_DESCRIPTION + "DESCRIPTION]\n" + "Example: " + COMMAND_WORD + " " + PREFIX_NAME + "Waiter "
-            + PREFIX_COMPANY + "Starbucks " + PREFIX_SALARY + "2500 " + PREFIX_REQUIREMENTS + "Hands and feet "
+    public static final String ENTITY_WORD = "job";
+    public static final String FULL_COMMAND = COMMAND_WORD + " " + ENTITY_WORD;
```
Comment (Quasant) [inline src/main/java/seedu/address/logic/commands/AddJobCommand.java:21]: Is this for purposes of testing?

File: src/main/java/seedu/address/logic/parser/AddContactCommandParser.java [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+                || !argMultimap.getPreamble().isEmpty()) {
+            throw new ParseException(String.format(MESSAGE_INVALID_COMMAND_FORMAT, AddContactCommand.MESSAGE_USAGE));
+        }
+
+        argMultimap.verifyNoDuplicatePrefixesFor(PREFIX_NAME, PREFIX_PHONE, PREFIX_EMAIL, PREFIX_ROLE);
+        Name name = ParserUtil.parseName(argMultimap.getValue(PREFIX_NAME).get());
+        Phone phone = ParserUtil.parsePhone(argMultimap.getValue(PREFIX_PHONE).get());
+        Email email = ParserUtil.parseEmail(argMultimap.getValue(PREFIX_EMAIL).get());
+        Role role = ParserUtil.parseRole(argMultimap.getValue(PREFIX_ROLE).get());
+        Set<Skill> skillList = ParserUtil.parseSkills(argMultimap.getAllValues(PREFIX_SKILL));
```
Comment (Quasant) [inline src/main/java/seedu/address/logic/parser/AddContactCommandParser.java:54]: I raised an observation about the `Skill` class in my later PR.

**Review by Quasant — APPROVED** [review]
LGTM

File: src/main/java/seedu/address/logic/commands/AddJobCommand.java (hunk 6 above)
Comment (KengHian) [reply src/main/java/seedu/address/logic/commands/AddJobCommand.java:21]: Partly yes, partly I just wanted to have a intermediate variable

File: src/main/java/seedu/address/logic/parser/AddContactCommandParser.java (hunk 7 above)
Comment (KengHian) [reply src/main/java/seedu/address/logic/parser/AddContactCommandParser.java:54]: Alright, then I will merge this version first.

**Pull request merged**